federation-js/src/service/printCoreSchema.ts

```diff
@@ -0,0 +1,581 @@
+
+  schema = new GraphQLSchema({
+    ...config,
+    directives: [
+      CoreDirective,
+      JoinFieldDirective,
+      JoinTypeDirective,
+      JoinOwnerDirective,
+      JoinGraphDirective,
+      ...config.directives,
+    ],
+    types: [FieldSetScalar, JoinGraphEnum, ...config.types],
+  });
+
+  const context: PrintingContext = {
+    sanitizedServiceNames,
+  }
+
+  return printFilteredSchema(
+    schema,
+    (n) => !isSpecifiedDirective(n),
+    isDefinedType,
+    options,
+    context,
+  );
+}
+
+export function printIntrospectionSchema(
+  schema: GraphQLSchema,
+  options?: Options,
+): string {
+  return printFilteredSchema(
+    schema,
+    isSpecifiedDirective,
+    isIntrospectionType,
+    options,
+  );
+}
+
+function isDefinedType(type: GraphQLNamedType): boolean {
+  return !isSpecifiedScalarType(type) && !isIntrospectionType(type);
+}
+
+function printFilteredSchema(
+  schema: GraphQLSchema,
+  directiveFilter: (type: GraphQLDirective) => boolean,
+  typeFilter: (type: GraphQLNamedType) => boolean,
+  options?: Options,
+  // Core addition - see `PrintingContext` type for details
+  context?: PrintingContext,
+): string {
+  const directives = schema.getDirectives().filter(directiveFilter);
+  const types = Object.values(schema.getTypeMap())
+    .sort((type1, type2) => type1.name.localeCompare(type2.name))
+    .filter(typeFilter);
+
+  return (
+    [printSchemaDefinition(schema)]
+      .concat(
+        directives.map((directive) => printDirective(directive, options)),
+        types.map((type) => printType(type, options, context)),
+      )
+      .filter(Boolean)
+      .join('\n\n') + '\n'
+  );
+}
+
+function printSchemaDefinition(schema: GraphQLSchema): string {
+  const operationTypes = [];
+
+  const queryType = schema.getQueryType();
+  if (queryType) {
+    operationTypes.push(`  query: ${queryType.name}`);
+  }
+
+  const mutationType = schema.getMutationType();
+  if (mutationType) {
+    operationTypes.push(`  mutation: ${mutationType.name}`);
+  }
+
+  const subscriptionType = schema.getSubscriptionType();
+  if (subscriptionType) {
+    operationTypes.push(`  subscription: ${subscriptionType.name}`);
+  }
+
+  return (
+    'schema' +
+    // Core change: print @core directive usages on schema node
+    printCoreDirectives() +
+    `\n{\n${operationTypes.join('\n')}\n}`
+  );
+}
+
+function printCoreDirectives() {
+  return [
+    'https://lib.apollo.dev/core/v0.1',
+    'https://lib.apollo.dev/join/v0.1',
+  ].map((feature) => `\n  @core(feature: ${printStringLiteral(feature)})`);
+}
+
+export function printType(
+  type: GraphQLNamedType,
+  options?: Options,
+  // Core addition - see `PrintingContext` type for details
+  context?: PrintingContext,
+): string {
+  if (isScalarType(type)) {
+    return printScalar(type, options);
+  } else if (isObjectType(type)) {
+    return printObject(type, options, context);
+  } else if (isInterfaceType(type)) {
+    return printInterface(type, options);
```

Comment:
I forgot to pass context here!
```suggestion
    return printInterface(type, options, context);
```

Reply:
Would it be feasible to make `context` a non-optional argument to the various functions, so this kind of bug can't happen?

Reply:
I guess I can comment here though, so: 25b5453